### PR TITLE
fix(hpExtension):Match the updated cs3d setPan/setZoom calls

### PR DESCRIPTION
**Pull request by wayfarer3130**
The CS3D options changed so that the entire display area/canvas center functionality is now embedded in with the setPan/setZoom calls rather than having separate options.

### CODE REVIEW

File: extensions/hp-extension/src/custom-viewport/setViewportZoomPan.ts [hunk 1]
```diff
@@ -5,30 +5,67 @@
  * @param value 
  * @param viewport 
  */
-const setViewportZoomPan = (id,value,viewport) => {
-  const {zoom,pan, flipHorizontal, flipVertical} = value;
-  if( zoom && zoom!==1 ) viewport.setZoom(zoom);
-  if( pan ) viewport.setPan(pan);
-  if( flipHorizontal || flipVertical ) {
+const setViewportZoomPan = (id, value, viewport) => {
```
Comment (wayfarer3130) [inline extensions/hp-extension/src/custom-viewport/setViewportZoomPan.ts:8]: Need to apply the imageCanvasPoint and initialDisplay area manually due to PR changes on CS3D

File: extensions/hp-extension/src/custom-viewport/setViewportZoomPan.ts [hunk 2]
```diff
@@ -5,30 +5,67 @@
...
+  if (initialDisplayArea) {
+    // TODO - use the correct calculation to actually fit things properly
+    // such that both initialDisplayArea's fit
+    initialZoom = 1 / Math.min(initialDisplayArea[0], initialDisplayArea[1]);
+    viewport.setZoom(initialZoom, true);
+  }
+
+  const imageData = viewport.getImageData();
+
+  if (imageCanvasPoint && imageData) {
```
Comment (wayfarer3130) [inline extensions/hp-extension/src/custom-viewport/setViewportZoomPan.ts:23]: This code works like the old version, but still has a few bugs - not sure why yet - seems like the canvas size isn't quite right, or the center point isn't actually the center of the image area.

File: extensions/hp-extension/src/hp/hpBreast.ts [hunk 3]
```diff
@@ -134,10 +137,9 @@ export default {
           viewportOptions: {
             toolGroupId: 'default',
             background: [127,127,0],
-            initialCenter: [0,0],
-            canvasCenter: [0,0],
             customViewportOptions: {
```
Comment (wayfarer3130) [inline extensions/hp-extension/src/hp/hpBreast.ts]: Needed to move everything into the zoomPan object

File: extensions/hp-extension/src/custom-viewport/setViewportZoomPan.ts [hunk 4]
```diff
@@ -5,30 +5,67 @@
...
  * @param viewport 
  */
-const setViewportZoomPan = (id,value,viewport) => {
-  const {zoom,pan, flipHorizontal, flipVertical} = value;
-  if( zoom && zoom!==1 ) viewport.setZoom(zoom);
-  if( pan ) viewport.setPan(pan);
-  if( flipHorizontal || flipVertical ) {
+const setViewportZoomPan = (id, value, viewport) => {
+  const { zoom, pan, flipHorizontal, flipVertical } = value;
+  const { imageCanvasPoint, initialDisplayArea } = value;
```
Comment (ladeirarodolfo) [inline extensions/hp-extension/src/custom-viewport/setViewportZoomPan.ts:10]: join imageCanvasPoint, initialDisplayArea  to { zoom, pan, flipHorizontal, flipVertical }

File: extensions/hp-extension/src/custom-viewport/setViewportZoomPan.ts [hunk 5]
```diff
@@ -5,30 +5,67 @@
...
+  if (imageCanvasPoint && imageData) {
+    const canvas = viewport.getCanvas();
+    const canvasX = imageCanvasPoint[2] ?? 0.5;
+    const canvasY = imageCanvasPoint[3] ?? 0.5;
+    const canvasPanX = canvas.width * (canvasX - 0.5);
+    const canvasPanY = canvas.height * (canvasY - 0.5);
+   
+    const canvasZero = viewport.worldToCanvas([0,0,0]);
+    const canvasEdge = viewport.worldToCanvas(imageData.dimensions);
+    const canvasImage = [canvasEdge[0]-canvasZero[0], canvasEdge[1]-canvasZero[1]];
```
Comment (ladeirarodolfo) [inline extensions/hp-extension/src/custom-viewport/setViewportZoomPan.ts:32]: there is no need on creating an array to later do a destructuring.

WE can just define imageWidth directly, no? const imageWidth = canvasEdge[0]-canvasZero[0];

File: extensions/hp-extension/src/custom-viewport/setViewportZoomPan.ts [hunk 6]
```diff
@@ -5,30 +5,67 @@
...
+  }
+
+  if (flipHorizontal || flipVertical) {
     viewport.flip(value);
   }
+
+  // Now set the sync value offsets
+  if (zoom && zoom !== 1) viewport.setZoom(zoom);
+  if (pan) {
+    console.log("Setting pan to", JSON.stringify(pan));
```
Comment (ladeirarodolfo) [inline extensions/hp-extension/src/custom-viewport/setViewportZoomPan.ts:46]: Do we really need to log this? Why not on Zoom?

File: extensions/hp-extension/src/custom-viewport/setViewportZoomPan.ts [hunk 7]
```diff
@@ -5,30 +5,67 @@
...
+    const canvasX = imageCanvasPoint[2] ?? 0.5;
+    const canvasY = imageCanvasPoint[3] ?? 0.5;
+    const canvasPanX = canvas.width * (canvasX - 0.5);
+    const canvasPanY = canvas.height * (canvasY - 0.5);
+   
+    const canvasZero = viewport.worldToCanvas([0,0,0]);
+    const canvasEdge = viewport.worldToCanvas(imageData.dimensions);
+    const canvasImage = [canvasEdge[0]-canvasZero[0], canvasEdge[1]-canvasZero[1]];
+    const [imgWidth,imgHeight] = canvasImage;
+    const imagePanX = imgWidth * (0.5-imageCanvasPoint[0]);
```
Comment (ladeirarodolfo) [inline extensions/hp-extension/src/custom-viewport/setViewportZoomPan.ts:34]: Ive seen some prettierfy incosistence here and some over this file. Is lint/prettier being applied at all?

File: extensions/hp-extension/src/hp/hpBreast.ts [hunk 8]
```diff
@@ -152,13 +154,12 @@ export default {
...
             background: [0,0,127],
-            initialCenter: [1,0],
-            canvasCenter: [1,0],
             customViewportOptions: {
               zoomPan: {
+                imageCanvasPoint: [1,0,1,0],
                 flipVertical: true,
               },
-            },
+            },
```
Comment (ladeirarodolfo) [inline extensions/hp-extension/src/hp/hpBreast.ts]: remove extra space please

**Review by ladeirarodolfo — CHANGES_REQUESTED** [review]
Thanks for your PR Bill,

I have added only minor cosmetic changes.
Please also ensure your linter/prettier runs over all changed files

File: extensions/hp-extension/src/custom-viewport/setViewportZoomPan.ts (hunk 7 above)
Comment (wayfarer3130) [reply extensions/hp-extension/src/custom-viewport/setViewportZoomPan.ts:34]: No, still need to setup lint appropriately on the files.  Will do it on master, and then apply the lint everywhere else.

File: extensions/hp-extension/src/synchronizers/initialZoomPanCallback.ts [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+  sViewport.setPan([panX, panY], true);
+
+  const { pan, zoom: previousZoom } = options;
+  // These are non-default initial values here, so don't pass true to reset
+  if( previousZoom ) sViewport.setZoom(previousZoom);
+  if( pan ) sViewport.setPan(pan);
+}
+
+export function storeCurrentZoomPan(synchronizerInstance: Synchronizer,  sourceViewport: Types.IViewportId) {
+  const renderingEngine = getRenderingEngine(sourceViewport.renderingEngineId);
```
Comment (ladeirarodolfo) [inline extensions/hp-extension/src/synchronizers/initialZoomPanCallback.ts:52]: This method is being tied to when ELEMENT is disabled. I'm wondering if getRenderingEngine and some other engine methods will work due to Element being ALREADY disabled

File: extensions/hp-extension/src/synchronizers/createSingleViewportSynchronizer.ts [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+
+function add(viewportInfo) {
+  const { viewportId } = viewportInfo;
+
+  this._viewportMap[viewportId] = { ...viewportInfo };
+
+  console.log("Add viewport info called for", viewportId);
+}
+
+function addEvent(eventName, listener) {
```
Comment (ladeirarodolfo) [inline extensions/hp-extension/src/synchronizers/createSingleViewportSynchronizer.ts:23]: I would recommend to implement also the removeEvent

File: extensions/hp-extension/src/synchronizers/createInitialZoomPanSynchronizer.ts [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+import createSingleViewportSynchronizer from './createSingleViewportSynchronizer';
+
+/**
+ * A helper that creates a new `Synchronizer` which listens to the `STACK_NEW_IMAGE`
+ * rendering event and calls the `initialZoomPanCallback`.
+ *
+ * @param synchronizerName - The name of the synchronizer.
+ * @returns A new `Synchronizer` instance.
+ */
+export default function createSynchronizer(
```
Comment (ladeirarodolfo) [inline extensions/hp-extension/src/synchronizers/createInitialZoomPanSynchronizer.ts:12]: do we have the ability to create, but what about destroy?

File: extensions/hp-extension/src/synchronizers/createInitialZoomPanSynchronizer.ts [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+  synchronizerName: string,
```
Comment (ladeirarodolfo) [inline extensions/hp-extension/src/synchronizers/createInitialZoomPanSynchronizer.ts:13]: What happens if by mistake we create a second synchroniser with same name?

File: extensions/hp-extension/src/index.tsx [hunk 13]
```diff
@@ -23,14 +24,16 @@ export default {
...
+    const { HangingProtocolService, SyncGroupService } = servicesManager.services;
+    // HangingProtocolService.addCustomViewportOption('zoomPan', "Set initial zoom, WITH camera event", 
+    //   setViewportZoomPan, 
+    //   storeViewportZoomPan);
     HangingProtocolService.addCustomAttribute('ViewCode', 'View Code Designator:Value', viewCodeAttribute);
     HangingProtocolService.addCustomAttribute('Laterality', 'Laterality of object', lateralityAttribute);
     HangingProtocolService.addCustomAttribute('seriesDescriptions', 'Series Descriptions', seriesDescriptionsFromDisplaySets);
     HangingProtocolService.addCustomAttribute('numberOfDisplaySets', 'Number of displays sets', numberOfDisplaySets);
+  
+    SyncGroupService.setSynchronizer('initialzoompan', initialZoomPan);
```
Comment (ladeirarodolfo) [inline extensions/hp-extension/src/index.tsx:36]: It seems there is a very specific version, that exposes servicesManager.services.SyncGroupService. setSynchronizer.
We again come to the coupling thing here. I was wondering if we could at least have a conditional to ensure there is this  method.

File: extensions/hp-extension/src/hp/hpHeart.ts [hunk 14]
```diff
@@ -245,11 +230,14 @@ export default {
...
           viewportOptions: {
             toolGroupId: 'default',
-            customViewportOptions: {
-              zoomPan: {
-                initialDisplayArea: [0.4, 0.4],
+            syncGroups: [
+              {
+                type: 'initialZoomPan',
+                options: { 
+                  initialDisplayArea: [0.4, 0.4],
```
Comment (ladeirarodolfo) [inline extensions/hp-extension/src/hp/hpHeart.ts:237]: there is a inconsistence on your linter. Some places like here there is a space. Some other there is not. Can you ensure your linter or project linter runs on all files?

**Review by ladeirarodolfo — CHANGES_REQUESTED** [review]

**Pull request merged**